### PR TITLE
Demo local S3 setup using localstack

### PR DESCRIPTION
**Pull request by motin**
Found this useful to try out the S3Aggregator locally without setting up S3 credentials etc. May also be useful in testing (https://github.com/mozilla/OpenWPM/issues/294)

### CODE REVIEW

**Comment by englehardt** [comment]
Very cool. it was unclear to me if this would work with s3fs, but it seems like it does.

**Comment by englehardt** [comment]
I have a WIP patch for #294 that does adds an option to save parquet datasets locally. I'll probably keep that around, but would like to incorporate localstack testing so it's end to end. Thanks for investigating this.

**Comment by motin** [comment]
> Very cool. it was unclear to me if this would work with s3fs, but it seems like it does.

Had to add a last commit to make it work with s3fs. Now this PR is ready for review :)

> I have a WIP patch for #294 that does adds an option to save parquet datasets locally. I'll probably keep that around, but would like to incorporate localstack testing so it's end to end. Thanks for investigating this.

Sounds good. Some tests should just just verify that the parquet part of the aggregation works, and end-to-end tests can incorporate s3.

**Comment by motin** [comment]
@englehardt Is there some auto-formatting command to use to fix the formatting issues reported by flask?

**Comment by englehardt** [comment]
@motin you can use a combination of `isort` and `autopep8` to fix errors automatically in-place. Here are some examples:

* `python3 -m isort -rc --diff .`
* `python -m autopep8 -r --diff .`

It looks like `autopep8` will fix errors [not flagged by `pep8`](https://github.com/hhatto/autopep8#features) (and thus not flagged by `flake8`, which is what we run for our tests). So the first run of that will change the docstrings on a bunch of methods. I'll make a PR that adds these changes.

**Comment by englehardt** [comment]
@motin Once #305 and #303 merge, can you rebase this on master?

File: test/test_s3_crawl.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    """ Runs a short test crawl.
+
+    This should be used to test any features that require real
+    crawl data. This should be avoided if possible, as controlled
+    tests will be easier to debug
+    """
+
+    def get_config(self, data_dir=""):
+        manager_params, browser_params = self.get_test_config(data_dir)
+        # Requires that `SERVICES=s3 localstack start` has been started
```
Comment (englehardt) [inline test/test_s3_crawl.py]: Does this happen automatically? Is this something we expect the users to run before they stat the tests? If so, can we automate as part of the test suite -- perhaps as a test fixture?

**Comment by eriktews** [comment]
Hi, would it possible to add it to TravisCI as well?

**Comment by motin** [comment]
> @motin Once #305 and #303 merge, can you rebase this on master?

Sure! They seem to be merged soon. Will wait and rebase.

**Comment by englehardt** [comment]
I'd like to get these tests set up so we can add additional tests for #323. I've gone ahead and finished the work here as part of #333.

File: test/test_s3_crawl.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from six.moves.urllib.parse import urlparse
+
+from ..automation import TaskManager
+from ..automation.utilities import db_utils, domain_utils
+from .openwpmtest import OpenWPMTest
+
+from .utilities import local_s3_bucket, download_s3_directory
+
+TEST_SITES = [
+    'http://example.com'
```
[the remark this answers is not in the document]
Comment (motin) [reply test/test_s3_crawl.py]: I see you resolved this in https://github.com/mozilla/OpenWPM/pull/333, thanks :)

File: test/test_s3_crawl.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    tests will be easier to debug
+    """
+
+    def get_config(self, data_dir=""):
+        manager_params, browser_params = self.get_test_config(data_dir)
+        # Requires that `SERVICES=s3 localstack start` has been started
+        manager_params['output_format'] = 's3'
+        manager_params['s3_bucket'] = local_s3_bucket()
+        manager_params['s3_directory'] = 'demo-local-s3'
+        browser_params[0]['profile_archive_dir'] =\
```
[the remark this answers is not in the document]
Comment (motin) [reply test/test_s3_crawl.py]: I see you resolved this in https://github.com/mozilla/OpenWPM/pull/333, thanks :)

File: test/test_s3_crawl.py (hunk 1 above)
Comment (motin) [reply test/test_s3_crawl.py]: I see you are resolving this in https://github.com/mozilla/OpenWPM/pull/333, thanks :)

File: test/test_s3_crawl.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        print("Copying the resulting S3 contents...")
+        destination = os.path.join(data_dir, 'local-s3-data')
+        download_s3_directory(
+            manager_params['s3_directory'],
+            destination,
+            manager_params['s3_bucket'])
+        print("Copied the resulting S3 contents to %s" % destination)
+
+        # TODO
+        assert True
```
[the remark this answers is not in the document]
Comment (motin) [reply test/test_s3_crawl.py]: I see you resolved this in https://github.com/mozilla/OpenWPM/pull/333, thanks :)

**Comment by motin** [comment]
> I'd like to get these tests set up so we can add additional tests for #323. I've gone ahead and finished the work here as part of #333.

Closing this PR in favor of #333

**Pull request closed without merging**